cpp/src/examples/babyllama/baby_llama_handler.cc

```diff
@@ -0,0 +1,269 @@
+    if (pos < num_elements - 1) {
+      // if we are still processing the input prompt, force the next prompt
+      // token
+      next = prompt_tokens[pos + 1];
+    } else {
+      // otherwise sample the next token from the logits
+      next = sample(&sampler, logits);
+    }
+    pos++;
+
+    torch::Tensor tensor = torch::tensor(next, torch::kLong);
+    tensor_vector.push_back(tensor);
+
+    // data-dependent terminating condition: the BOS (=1) token delimits
+    // sequences
+    if (next == 1) {
+      break;
+    }
+
+    token = next;
+
+    // init the timer here because the first iteration can be slower
+    if (start == 0) {
+      start = time_in_ms();
+    }
+  }
+
+  // report achieved tok/s (pos-1 because the timer starts after first
+  // iteration)
+  if (pos > 1) {
+    long end = time_in_ms();
+    auto token_per_sec = (pos - 1) / (double)(end - start) * 1000;
+    std::cout << "Achieved tok per sec: " << token_per_sec << std::endl;
+  }
+
+  delete[] prompt_tokens;
+
+  torch::Tensor stacked_tensor = torch::stack(tensor_vector);
+  return stacked_tensor;
+}
+
+void LlmHandler::Postprocess(
+    const torch::Tensor& data,
+    std::pair<std::string&, std::map<uint8_t, std::string>&>& idx_to_req_id,
+    std::shared_ptr<torchserve::InferenceResponseBatch>& response_batch) {
+  for (const auto& kv : idx_to_req_id.second) {
+    try {
+      int64_t num_elements = data.numel();
+      auto data_ptr = data.data_ptr<int64_t>();
+      int64_t token = 1;
+      std::string concatenated_string;
+      for (int64_t i = 0; i < num_elements; ++i) {
+        char* piece = decode(&tokenizer, token, data_ptr[i]);
+        std::string piece_string(piece);
+        token = data_ptr[i];
+        concatenated_string += piece_string;
+      }
+
+      std::cout << "Concatenated String: " << concatenated_string << std::endl;
+
+      auto response = (*response_batch)[kv.second];
+
+      response->SetResponse(200, "data_type",
+                            torchserve::PayloadType::kDATA_TYPE_STRING,
+                            concatenated_string);
+    } catch (const std::runtime_error& e) {
+      TS_LOGF(ERROR, "Failed to load tensor for request id: {}, error: {}",
+              kv.second, e.what());
+      auto response = (*response_batch)[kv.second];
+      response->SetResponse(500, "data_type",
+                            torchserve::PayloadType::kDATA_TYPE_STRING,
+                            "runtime_error, failed to postprocess tensor");
+    } catch (const c10::Error& e) {
+      TS_LOGF(ERROR,
+              "Failed to postprocess tensor for request id: {}, error: {}",
+              kv.second, e.msg());
+      auto response = (*response_batch)[kv.second];
+      response->SetResponse(500, "data_type",
+                            torchserve::PayloadType::kDATA_TYPE_STRING,
+                            "c10 error, failed to postprocess tensor");
+    }
+  }
+
+  free_sampler(&sampler);
```

Comment:
If we free the sample after postprocessing a batch, the next batch will fail. Better place should be the class destructor.

Reply:
Moved it inside the class destructor